com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveJavaExecutableHandler.java

```diff
@@ -0,0 +1,101 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package com.microsoft.java.debug.plugin.internal;
+
+import java.io.File;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.launching.IVMInstall;
+import org.eclipse.jdt.launching.JavaRuntime;
+
+import com.microsoft.java.debug.core.Configuration;
+
+public class ResolveJavaExecutableHandler {
+    private static final Logger logger = Logger.getLogger(Configuration.LOGGER_NAME);
+    private static final String[] candidateJavaExecs = {
+        "java",
+        "java.exe",
+        "javaw",
+        "javaw.exe",
+        "j9",
+        "j9.exe",
+        "j9w",
+        "j9w.exe"
+    };
+    private static final String[] candidateJavaBins = {
+        File.separator,
+        "bin" + File.separatorChar,
+        "jre" + File.separatorChar + "bin" + File.separatorChar
+    };
+
+    /**
+     * Resolve the java executable path from the project's java runtime.
+     */
+    public static String resolveJavaExecutable(List<Object> arguments) throws Exception {
+        try {
+            String mainClass = (String) arguments.get(0);
+            String projectName = (String) arguments.get(1);
+            IJavaProject targetProject = null;
+            if (StringUtils.isNotBlank(projectName)) {
+                targetProject = JdtUtils.getJavaProject(projectName);
+            } else {
```

Comment:
If a project is not specified, should we just use whatever JVM that the debugger is using to launch the app?

Reply:
In launch.json, `project` info is optional. In the case of undefined, the debugger will try to resolve the project info from the main class. See the `else` branch.